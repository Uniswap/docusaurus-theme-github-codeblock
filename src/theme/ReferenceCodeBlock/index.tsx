import React, { useReducer } from 'react'
import CodeBlock from '@theme-init/CodeBlock'

import type { ReferenceCodeBlockProps, GitHubReference, DispatchMessage } from '../types'

const initialFetchResultState = {
    code: 'loading...',
    error: null,
    loading: null,
}

/**
 * parses GitHub reference
 * @param {string} ref url to github file
 */
function parseReference (ref: string): GitHubReference {
    const fullUrl = ref.slice(ref.indexOf('https'), -1)
    const [url, loc] = fullUrl.split('#')
    const [org, repo, blob, branch, ...pathSeg] = new URL(url).pathname.split('/').slice(1)
    const [fromLine, toLine] = loc
        ? loc.split('-').map((lineNr) => parseInt(lineNr.slice(1), 10))
        : [0, Infinity]

    return {
        url: `https://raw.githubusercontent.com/${org}/${repo}/${branch}/${pathSeg.join('/')}`,
        fromLine,
        toLine,
        title: pathSeg.join('/')
    }
}

async function fetchCode ({ url, fromLine, toLine }: GitHubReference, fetchResultStateDispatcher: React.Dispatch<DispatchMessage>) {
    let res: Response

    try {
        res = await fetch(url)
    } catch (err) {
        return fetchResultStateDispatcher({ type: 'error', value: err })
    }

    if (res.status !== 200) {
        const error = await res.text()
        return fetchResultStateDispatcher({ type: 'error', value: error })
    }

    const body = (await res.text()).split('\n').slice(fromLine, (toLine || fromLine) + 1)
    const preceedingSpace = body.reduce((prev: number, line: string) => {
        const spaces = line.match(/^\s+/)

        if (spaces) {
            return Math.min(prev, spaces[0].length)
        }

        return 0
    }, Infinity)

    return fetchResultStateDispatcher({
        type: 'loaded',
        value: body.map((line) => line.slice(preceedingSpace)).join('\n')
    })
}

function ReferenceCode(props: ReferenceCodeBlockProps) {
    const [fetchResultState, fetchResultStateDispatcher] = useReducer(
        (prevState: any, { type, value }: DispatchMessage) => {
            switch (type) {
                case 'reset': {
                return initialFetchResultState;
                }
                case 'loading': {
                return {...prevState, loading: true};
                }
                case 'loaded': {
                return {...prevState, code: value, loading: false};
                }
                case 'error': {
                return {...prevState, error: value, loading: false};
                }
                default:
                return prevState;
            }
        },
        initialFetchResultState,
    )

    console.log(props.children);

    const codeSnippetDetails = parseReference(props.children)
    fetchCode(codeSnippetDetails, fetchResultStateDispatcher)
    const customProps = {
        ...props,
        metastring: ` title="${codeSnippetDetails.title}"`,
        children: initialFetchResultState.code
    }

    return (
        <div>
            <CodeBlock {...customProps}>{fetchResultState.code}</CodeBlock>
        </div>
    );
}

export default ReferenceCode;
