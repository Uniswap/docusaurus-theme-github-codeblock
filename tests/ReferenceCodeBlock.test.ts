import { timeStamp } from 'console'
import { parseReference } from '../src/theme/ReferenceCodeBlock/index'

test('should parse GitHub reference properly', () => {
    expect(parseReference('https://github.com/saucelabs/docusaurus-theme-github-codeblock/blob/main/src/theme/ReferenceCodeBlock/index.tsx'))
        .toMatchSnapshot()
    expect(parseReference('https://github.com/saucelabs/docusaurus-theme-github-codeblock/blob/main/src/theme/ReferenceCodeBlock/index.tsx#L105-L108'))
        .toMatchSnapshot()
})
