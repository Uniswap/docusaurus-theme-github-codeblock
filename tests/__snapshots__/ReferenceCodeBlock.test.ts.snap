// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse GitHub reference properly 1`] = `
Object {
  "fromLine": 0,
  "title": "src/theme/ReferenceCodeBlock/index.ts",
  "toLine": Infinity,
  "url": "https://raw.githubusercontent.com/saucelabs/docusaurus-theme-github-codeblock/main/src/theme/ReferenceCodeBlock/index.ts",
}
`;

exports[`should parse GitHub reference properly 2`] = `
Object {
  "fromLine": 104,
  "title": "src/theme/ReferenceCodeBlock/index.tsx",
  "toLine": 9,
  "url": "https://raw.githubusercontent.com/saucelabs/docusaurus-theme-github-codeblock/main/src/theme/ReferenceCodeBlock/index.tsx",
}
`;
